### DISCUSSION TITLE
How to get elements of a ListKind variable's value?

### DISCUSSION THREAD

**Discussion by zzxwill** [cuelang/cue#600] [General] [answered]
In `test.cue`, there are two variables `cmd` and `rules`. I'd like to get all elements of their values. For `cmd`, to get type `string`. For `rules`, to get element `path`, `rewriteTarget` and their types.
```
cmd: [...string]

rules: [...{
	path:          string
	rewriteTarget: *"" | string
}]
```

Here is my code, but I just could not even get the values, nor their elements.

```
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"

	"cuelang.org/go/cue"
)

func main() {
	path, _ := filepath.Abs("./test.cue")
	data, err := ioutil.ReadFile(path)

	if err != nil {
		fmt.Println(err)
		os.Exit(-1)
	}
	r := cue.Runtime{}
	template, _ := r.Compile("", string(data))
	stc, _ := template.Value().Struct()
	for i := 0; i < stc.Len(); i++ {
		val := stc.Field(i).Value
		switch val.IncompleteKind() {
		case cue.ListKind:
			iter, _ := val.List()
			for iter.Next() {
				fmt.Println(iter.Value())
			}

		}
	}
}

```

I also put the .cue file and the go file at https://github.com/zzxwill/try-cloudnative/blob/master/poc.go. Please help take a look at this question. Thanks.

**Comment by mpvl** [comment, marked as answer]
In CUE, currently the only way to get these values are doing something like:
```
cmd: [...string]

type: (cmd & [_])[0]
```
Which results in `string` for `type`

In the API you can use `Value.Elem()` to get the same.

**Comment by mpvl** [comment]
We are considering allowing `?` to do a lookup op optional fields. So for the following struct
```
a: {
    opt?: string
}
```
one could write `a.opt?` to get the value string. Similarly, for a struct
```
a: [string]: <100
a: [>"e"]:  >10
```
for `a.foo?` one would get `>10 & <100`, that is the value that would be given when adding `foo` as a field.

Similarly then, and assuming we will support using a selector for numeric values, one could do `cmd.0?` to query what the first element would be if it were defined. So the get the same for the "next" undefined entry, one could query `cmd[len(cmd)]?`.

Anyway, that is just a thought. But this seems to come up often enough to warrant it. People can weigh in here if they think it is useful.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/600.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
